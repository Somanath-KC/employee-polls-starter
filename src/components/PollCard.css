/* PollCard Component Styles */

.poll-card-header {
  display: flex;
  align-items: center;
  gap: 12px;
  margin-bottom: 16px;
}

.poll-card-avatar {
  width: 40px;
  height: 40px;
}

.poll-card-author-name {
  font-weight: 600;
  margin: 0;
  color: #212529;
}

.poll-card-timestamp {
  font-size: 14px;
  margin: 0;
}

.poll-card-title {
  font-size: 18px;
  font-weight: 500;
  color: #212529;
  margin-bottom: 12px;
}

.poll-card-options {
  margin-bottom: 16px;
}

.poll-card-option {
  color: #6c757d;
  margin: 8px 0;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.poll-card-option-label {
  font-weight: 500;
}

.poll-card-footer {
  display: flex;
  align-items: center;
  justify-content: space-between;
}

.poll-card-status {
  display: inline-flex;
  align-items: center;
  padding: 4px 12px;
  border-radius: 12px;
  font-size: 12px;
  font-weight: 500;
}

.poll-card-status.answered {
  background-color: #d4edda;
  color: #155724;
  border: 1px solid #c3e6cb;
}

.poll-card-status.not-answered {
  background-color: #fff3cd;
  color: #856404;
  border: 1px solid #ffeaa7;
}

.poll-card-view-btn {
  text-decoration: none;
  font-size: 14px;
  background-color: #e7f3ff;
  color: #007bff;
  border: 1px solid #b3d9ff;
}

.poll-card-view-btn:hover {
  background-color: #cce7ff;
  text-decoration: none;
}