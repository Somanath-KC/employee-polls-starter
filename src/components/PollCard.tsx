import { Link } from "react-router";
import { useAppSelector } from "../store/hooks";
import type { Question } from "../utils/_DATA";

interface PollCardProps {
  question: Question;
}

export default function PollCard({ question }: PollCardProps) {
  const users = useAppSelector((state) => state.users.users);
  const currentUser = useAppSelector((state) => state.auth.user);

  const author = users[question.author];
  const hasAnswered =
    (currentUser && question.optionOne.votes.includes(currentUser)) ||
    (currentUser && question.optionTwo.votes.includes(currentUser));

  const formatDate = (timestamp: number) => {
    return new Date(timestamp).toLocaleDateString("en-US", {
      year: "numeric",
      month: "long",
      day: "numeric",
      hour: "2-digit",
      minute: "2-digit",
    });
  };

  return (
    <div className="bg-white rounded-lg shadow-md p-6 hover:shadow-lg transition-shadow">
      <div className="flex items-center space-x-3 mb-4">
        {author && (
          <>
            <img
              src={author.avatarURL}
              alt={author.name}
              className="w-10 h-10 rounded-full"
            />
            <div>
              <p className="font-semibold text-gray-900">{author.name}</p>
              <p className="text-sm text-gray-500">
                {formatDate(question.timestamp)}
              </p>
            </div>
          </>
        )}
      </div>

      <h3 className="text-lg font-medium text-gray-900 mb-3">
        Would You Rather...
      </h3>

      <div className="space-y-2 mb-4">
        <p className="text-gray-600 truncate">
          <span className="font-medium">A:</span> {question.optionOne.text}
        </p>
        <p className="text-gray-600 truncate">
          <span className="font-medium">B:</span> {question.optionTwo.text}
        </p>
      </div>

      <div className="flex items-center justify-between">
        <span
          className={`inline-flex items-center px-2.5 py-0.5 rounded-full text-xs font-medium ${
            hasAnswered
              ? "bg-green-100 text-green-800"
              : "bg-yellow-100 text-yellow-800"
          }`}
        >
          {hasAnswered ? "Answered" : "Not Answered"}
        </span>

        <Link
          to={`/questions/${question.id}`}
          className="inline-flex items-center px-4 py-2 border border-transparent text-sm font-medium rounded-md text-blue-600 bg-blue-100 hover:bg-blue-200 transition-colors"
        >
          View Poll
        </Link>
      </div>
    </div>
  );
}
