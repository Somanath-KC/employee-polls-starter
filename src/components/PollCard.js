import React from "react";
import { Link } from "react-router-dom";
import { useAppSelector } from "../store/hooks";
import "./PollCard.css";

export default function PollCard({ question }) {
  const users = useAppSelector((state) => state.users.users);
  const currentUser = useAppSelector((state) => state.auth.user);

  const author = users[question.author];
  const hasAnswered =
    (currentUser && question.optionOne.votes.includes(currentUser)) ||
    (currentUser && question.optionTwo.votes.includes(currentUser));

  const formatDate = (timestamp) => {
    return new Date(timestamp).toLocaleDateString("en-US", {
      year: "numeric",
      month: "long",
      day: "numeric",
      hour: "2-digit",
      minute: "2-digit",
    });
  };

  return (
    <div className="poll-card">
      <div className="poll-card-header">
        {author && (
          <>
            <img
              src={author.avatarURL}
              alt={author.name}
              className="avatar poll-card-avatar"
            />
            <div>
              <p className="poll-card-author-name">{author.name}</p>
              <p className="text-muted poll-card-timestamp">
                {formatDate(question.timestamp)}
              </p>
            </div>
          </>
        )}
      </div>

      <h3 className="poll-card-title">Would You Rather...</h3>

      <div className="poll-card-options">
        <p className="poll-card-option">
          <span className="poll-card-option-label">A:</span>{" "}
          {question.optionOne.text}
        </p>
        <p className="poll-card-option">
          <span className="poll-card-option-label">B:</span>{" "}
          {question.optionTwo.text}
        </p>
      </div>

      <div className="poll-card-footer">
        <span
          className={`poll-card-status ${
            hasAnswered ? "answered" : "not-answered"
          }`}
        >
          {hasAnswered ? "Answered" : "Not Answered"}
        </span>

        <Link
          to={`/questions/${question.id}`}
          className="btn poll-card-view-btn"
        >
          View Poll
        </Link>
      </div>
    </div>
  );
}
