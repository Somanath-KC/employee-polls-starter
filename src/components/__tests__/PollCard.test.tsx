import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router";
import { configureStore } from "@reduxjs/toolkit";
import authReducer from "../../features/auth/authSlice";
import usersReducer from "../../features/users/usersSlice";
import questionsReducer from "../../features/questions/questionsSlice";
import PollCard from "../PollCard";

const mockQuestion = {
  id: "test-question",
  author: "testuser",
  timestamp: 1609459200000, // Jan 1, 2021
  optionOne: {
    votes: ["user1"],
    text: "Option One Text",
  },
  optionTwo: {
    votes: ["user2", "user3"],
    text: "Option Two Text",
  },
};

const mockUsers = {
  testuser: {
    id: "testuser",
    name: "Test User",
    avatarURL: "https://example.com/avatar.jpg",
    answers: {},
    questions: ["test-question"],
    password: "password",
  },
};

const createMockStore = (authUser: string | null = null) => {
  return configureStore({
    reducer: {
      auth: authReducer,
      users: usersReducer,
      questions: questionsReducer,
    },
    preloadedState: {
      auth: {
        user: authUser,
        isAuthenticated: !!authUser,
      },
      users: {
        users: mockUsers,
        loading: false,
        error: null,
      },
      questions: {
        questions: {},
        loading: false,
        error: null,
      },
    },
  });
};

// Test 5: Snapshot test for PollCard component
test("PollCard component renders correctly", () => {
  const store = createMockStore("testuser");

  const { container } = render(
    <MemoryRouter>
      <Provider store={store}>
        <PollCard question={mockQuestion} />
      </Provider>
    </MemoryRouter>
  );

  expect(container.firstChild).toMatchSnapshot();
});
