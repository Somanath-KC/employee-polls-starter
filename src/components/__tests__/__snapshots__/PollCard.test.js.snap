// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PollCard component renders correctly 1`] = `
<div
  class="poll-card"
>
  <div
    class="poll-card-header"
  >
    <img
      alt="Test User"
      class="avatar poll-card-avatar"
      src="https://example.com/avatar.jpg"
    />
    <div>
      <p
        class="poll-card-author-name"
      >
        Test User
      </p>
      <p
        class="text-muted poll-card-timestamp"
      >
        January 1, 2021 at 12:00 AM
      </p>
    </div>
  </div>
  <h3
    class="poll-card-title"
  >
    Would You Rather...
  </h3>
  <div
    class="poll-card-options"
  >
    <p
      class="poll-card-option"
    >
      <span
        class="poll-card-option-label"
      >
        A:
      </span>
       
      Option One Text
    </p>
    <p
      class="poll-card-option"
    >
      <span
        class="poll-card-option-label"
      >
        B:
      </span>
       
      Option Two Text
    </p>
  </div>
  <div
    class="poll-card-footer"
  >
    <span
      class="poll-card-status not-answered"
    >
      Not Answered
    </span>
    <a
      class="btn poll-card-view-btn"
      href="/questions/test-question"
    >
      View Poll
    </a>
  </div>
</div>
`;
